### ISSUE TITLE
Documentation Issue

### ISSUE THREAD

**Issue by stone212** [sammy007/open-ethereum-pool#208]
This is an issue regarding the documentation.  An earlier similar issue was accidentally closed without the documentation being updated or the questions in this ticket answered.  Since documentation is an essential part of any project, I'm happy to help if I can.

I will post the confusing text and then ask questions. 

```
Configure nginx to serve API on /api subdirectory. Configure nginx to serve www/dist as static website.
Serving API using nginx

Create an upstream for API:

upstream api {
    server 127.0.0.1:8080;
}

and add this setting after location /:

location /api {
    proxy_pass http://api;
}
```

First, it is not clear if the code ("upstream api" and such) is supposed to be for nginx.  Is it?  That is the first thing to make clear.

If so, what file does it go into?  Not everyone is an nginx expert, and also nginx has many ways to configure it.  If the code is not for nginx, where is it for?

Second, please put an explanation in the doc about the difference between the static website and the API.  What do each of them do? **This is really important and there are other questions about this.**

Is there a specific reason to run nginx?  Could an apache2 reverse proxy work instead?

Are you saying that there should be two websites?  one at "domain.com" and one at "domain.com/api" or what?

**Comment by mikeyb** [comment]
1. Yes, read nginx docs, they are wonderful
2. 1 will solve confusion on 2

Yes, nginx is way way faster than apache.  Apache is like driving a semi down a bike trail here.

No.  It is 1 website using a path of domain.com/api which feeds an nginx upstream of a locally running process listening on port 8080.  Refer to answer 1

**Comment by sammy007** [comment]
I agree with @mikeyb.

**Comment by stone212** [comment]
@mikeyb That does answer the question, thank you.  But I'll use Apache2 unless I find that the code is fundamentally incompatible.  Apache2 is more like a swiss army knife than the precise tool, but once you know your swiss army knife really well, you grab for it first.  It's not like this pool the only thing I have running on the server, and anyway I find Apache2 to be faster for dynamic content, although I'm not sure about golang?

I might try nginx but if not then I'll submit my Apache2 howto so you can add it to the wiki once I figure it out.

**Comment by mikeyb** [comment]
nginx is more the swiss army knife imo.  apache is more a giant boat with the same tools.  from decades working in web hosting almost exclusively with apache I now turn to nginx first and can't think of a time in the past couple years apache fit the requirements better.

**Comment by stone212** [comment]
@mikeyb How does nginx handle running a go program?  Does it let go run its own web server or does it have some sort of module to handle go?

**Comment by mikeyb** [comment]
nginx has many uses, in the case of using a golang webserver, nginx would act as proxy between the browser and go server. nginx would take the request, forward it to the upstream (go server), get the response back from the backend and forward it back to the user.

**Comment by stone212** [comment]
@mikeyb That's good to know, but I generally do those things with Apache2.  Part of the purpose of my help here is to assist you in separating the open-ethereum-pool project from nginx-specific documentation.  Any project that depends on any one web service (even if it is your favorite) is not very well developed, and since I think this project has potential I would like to help make it less niche-based.  

Debating the value of nginx vs. apache2 is not part of this ticket.  It's a silly little geek argument.  The point is that open-ethereum-pool has to be independent and I opened this ticket because I am open to helping with the documentation.  If you can clear up the docs, including the nginx docs, then I will know what is the nginx-specific stuff vs. the open-ethereum-pool stuff.  Then I can know what is specific to the code, and I can get it working with apache2.  Then I can write a little "how to configure with apache2" entry for the wiki.

The benefits of community collaboration... but it does require collaboration on the part of the development team.

**Comment by sammy007** [comment]
tl;dr: you must serve `dist` directory with static website and proxy pool api. You can use any software you want for that. COnfiguration of nginx, apache, haproxy, etc is out of scope of this project and issue tracker.

**Comment by stone212** [comment]
@sammy007  Thank you for replying.  

> COnfiguration of nginx, apache, haproxy, etc is out of scope of this project

Agreed 100%  I do not need help with anything Apache2 (etc.) related.  The issue here is with English, honestly.  You said:

"you must serve dist directory with static website and proxy pool api."  In English, this makes no sense.  If taken literally, it means 'the dist directory must be served [as] both a static website and simultaneously be served as a proxy pool api' but this is not possible.  So maybe you said that dist needs to be a static website, and also proxy pool api needs to be a static website?  But what is "proxy pool api" in this case and if it is the API, then how can that be a static website?

Do you see the problem?

But actually, I think you are saying two things:

1 - You must serve dist directory [as] a static website

2 - ?????

I simply do not know what you mean by "proxy pool api" or what you are saying to do with it.  It would also help if you could give a TL;DR on what the /dist static website does and what the API does.  This is not clear until it is running, but making it clear might (?) help me know how to make it run.

**Comment by hashmonkey** [comment]
you simply need to use apache2 reverse proxy then to serve /api then, and put the static dist files on Root path (ex: /var/www/html or whatever you have). 

some rtfm'ing at: https://u6256229.ct.sendgrid.net/wf/click?upn=D2O9UnPOsmzrLVN4IYOpyzpW1kKkcvJ026zKHVZ5T-2BI9hAMCMMoBr-2BZQG7EzhOXSn0ZmiENMd1QWIx1VCjIm0Bs6OGvfuqiTf6gKJdBEx-2BLEt5JqZteU-2BW7-2BC-2FWTSt-2BFd2efEmXtdvL9ILBLRJ7YLLEjdMRu38abUiGBCQcOSIEO0v0okdYuLFS2NhdvowyJ_s5UH8bhnBaopYjthcNSRcZSbK0lQpcsdiSzd48pkjGDHFpb3RfHfYXlZsPyWG-2Fse0uQx5-2BofUrtl2AF3s44AIGy9LWrrzUyW21WLhet-2BgGwbMRLbtjAYT40iKMco9P3WlHCjP6ERinArfZwcs3MmnE8SIUXWHYXaXSPm5u87IoKApzwcpMqpdYK6xQI8hHEI6LQoTe2-2BcK-2BW9iCHSjJFZ5hcWzVIYhWzmyQBN7VlGy-2BV7x-2Fm6hq8l2m2p1XsgrQxDHI5yVPmASbmAdqfzEu8-2Fne4zHEJu7QqftIriTJMX8RwDbBBrcpBOknPjstuNUP4D8Y79RIL39PMJQwR-2F0OvxA-3D-3D

**Comment by stone212** [comment]
@hashmonkey Thank you for replying but you misinterpreted the Issue.  I do not need help working with reverse proxies, especially in Apache2.

This ticket is regarding confusing documentation, not a technical issue.  Your link is out of scope for the issue/ticket, as the ticket has absolutely zero relation to reverse proxy configuration.  The ticket regards the confusing documentation.

**Comment by stone212** [comment]
@hashmonkey @sammy007 

Maybe this will help.  This ticket has two parts.  The first is solved by removing this:
```

Configure nginx to serve API on /api subdirectory. Configure nginx to serve www/dist as static website.
Serving API using nginx

Create an upstream for API:

upstream api {
    server 127.0.0.1:8080;
}

and add this setting after location /:

location /api {
    proxy_pass http://api;
}
```

with this:

> Configure your web server to serve the /api directory as <hostname>/api, and to separately serve the static site in /dist as the document root.  

Once you do that, you make the process sooooo much more clear.  Or actually: less confusing.  Because the problem was not the code.  The problem was that the documentation was confusing.

Then there is the part that is just as important.  Pasted from my original:

> Second, please put an explanation in the doc about the difference between the static website and the API. What do each of them do? This is really important and there are other questions about this.

Since I have not successfully gotten the pool to work, I don't know the answer the the above.  But now that I know what sites are to be served, I can do that correctly and I will know that whatever issues I am having are NOT with the reverse proxy.  (Actually I already knew that, because I am using the build-in development server and the API still does not show anything :( - but that's a different ticket)

**Comment by mikeyb** [comment]
Umm, if you remove the api web server configuration, you will not have a working pool.  That step is needed regardless if you think it belongs or not.  This pool takes 15 minutes to setup.

**Comment by mikeyb** [comment]
> But now that I know what sites are to be served, I can do that correctly and I will know that whatever issues I am having are NOT with the reverse proxy.

WRONG!  no api, no communications.  In over your head still

**Comment by stone212** [comment]
@mikeyb said:

> Umm, if you remove the api web server configuration, you will not have a working pool.

Who said anything about removing the api web server configuration?  Anyway, the pool is working fine except for 3rd party payouts.  Are you saying that having the web api running is connected to 3rd party payouts?

I appreciate that you responded but half-replies are not helpful.  Words like "working" are vague.

> This pool takes 15 minutes to setup.

I am sure that that is true... once it is documented and the Parity payout bugs are fixed.

> WRONG! no api, no communications.

Interesting.  When you say "no communications" what are you referring to?  Communications between what and what?  This is not mentioned in the documentation.

**Comment by sammy007** [comment]
> working fine except for 3rd party payouts

What issues you are facing with 3rd party payouts? Pool process payouts from matured mined funds proportionally. They only caveat is that you have to send some dust for tx fees to the wallet because pool does not reserve tx fees and credit all block rewards by default. There are various options to keep tx fees from block reward to cover expenses.

**Comment by sammy007** [comment]
This pool as mentioned requires 15 minutes to set up and configure, if somebody discovered ethereum 3 days ago it indeed requires some time spent on education, and this is really good because if somebody wanna play with other people expensive investments (hashrate) this someone must have some experience, for profit or for non-profit it does not matter, people can lose money if they put their investments on amateur service.

**Comment by sammy007** [comment]
Good this documentation or not, it's relative. Pool is published under terms of devgrants, I am definitely not going to waste my time on it more, I lost approximately $20+K because users refused to follow this https://github.com/sammy007/open-ethereum-pool/issues/27 so i can't claim my fair reward from ETH Foundation for supporting this project.

**Comment by stone212** [comment]
@sammy007 That's terrible about the extradata.  Really no excuse for using something and not fully participating.  If I ever get this running, I will set extradata.

> This pool as mentioned requires 15 minutes to set up and configure

You keep saying that.  But saying that does not make it true.  However, I did manage to get the UI working today.  (It is very nice BTW!)  It was simple, but required weeks of trial and error because the doc did not mention many important things.  I still have not made the website public but this is going to be very easy and anyway, it is not in scope of the project.  Web servers are easy.  Getting the dist to talk to api was easy.  Getting a straight answer on what dist and api are?  That took weeks.

> Good this documentation or not, it's relative.

When I am finished I will send you new docs.  You can decide if you want to use them.

> What issues you are facing with 3rd party payouts? Pool process payouts from matured mined funds proportionally.

The issue is that no payouts are happening at all.  Only the address of the "author" on the mining node is receiving any Ether.  I don't know if there is anything for me to configure.  I already have "threshold" set to 50 but this should not matter because I am on a private net and have 25000+ FakeETH.  I will investigate as soon as I have time.  I will start with this, but it is very unclear:
https://github.com/sammy007/open-ethereum-pool/blob/master/docs/PAYOUTS.md

**Comment by sammy007** [comment]
> If I ever get this running, I will set extradata.

No longer makes sense.

> The issue is that no payouts are happening at all

As long as payout node has private key of coinbase address and unlocked they must go.
The `address` https://github.com/sammy007/open-ethereum-pool/blob/master/config.example.json#L102 is a coinbase address because eth requires a source address to be specified. If it does not there must be an error.

Yes the Payouts.md is not clear because it's complicated and my english sometimes good and sometimes it's terrible, depends on how I am tired.

**Comment by stone212** [comment]
@sammy007 

> As long as payout node has private key of coinbase address and unlocked they must go.

What do you mean by "payout node"?  I do have a node that has an unlocked address.  And I have a separate miner that is configured to pay to the unlocked address on this other node.  Is this what you mean?

> The address https://github.com/sammy007/open-ethereum-pool/blob/master/config.example.json#L102 

My config.json has "0x0" as the "address" on line 102.  That is because we do not have one single payout address.  Every miner that connects to the pool has a different payout address because every miner is owned by a different person.

> is a coinbase address because eth requires a source address to be specified. If it does not there must be an error.

I don't understand these sentences.  What do you mean by "source address"?  do you mean that line 102 should have NOT the payout address but actually have the "author" address from the mining pool node?  NOT the unlocked address for the miner.  YES the author address from the mining pool node?  (I will try this now).

**Comment by stone212** [comment]
@sammy007 

I have written some documentation.  If you would like to create a docs/WebUIHowto.md file, I can try to submit a pull request.  (Haven't done that with GitHub before, just regular git but I guess it's not much harder?)

Or I can post here.  It does not include reverse proxy yet as I have not worked on that.  coming soon.

**Comment by stone212** [comment]
@sammy007 

> Yes the Payouts.md is not clear because it's complicated and my english sometimes good and sometimes it's terrible, depends on how I am tired.

I'm sure your English is better than I would be if I tried to speak your language.  Anyway this is part of why I am offering to help.  For example, I know what you mean by "depends on how I am tired" but if that was technical terms, the difference between "depends on how I am tired" and "depends on how tired I am" can be a completely different configuration.  And it is these explanations that I am trying to understand.

For example, the ember question: your answer makes perfect sense.  It is what seemed logical.  But it was not what was written, so I wanted clarity.

**Comment by AndreyNag** [comment]
Hi. Tell me please where i have to insert 
`upstream api {
    server 127.0.0.1:8080;
}` and etc?
Thx.

**Comment by sammy007** [comment]
https://nginx.ru/ru/docs/http/ngx_http_upstream_module.html
